docs/developer/python-openflow.rst

```diff
@@ -0,0 +1,206 @@
+
+packing
++++++++
+To pack a pyof object, simply call its ``pack()`` method, which returns a
+``bytes`` object of the binary representation.
+
+unpacking
++++++++++
+To unpack a buffer as pyof object, simply initialize a new object and
+call its ``unpack()`` method, passing the buffer as the argument. The
+buffer will be unpacked in place setting the attributes of the object
+instance used to make the call.
+
+
+New definitions
+===============
+``python-openflow`` provides some base classes and basic type classes meant to
+be used in a new message/struct class definition.
+
+Difference between "structs" and "messages"
++++++++++++++++++++++++++++++++++++++++++++
+Messages are GenericStruct derivatives who carry a header attribute containing
+an OpenFlow header struct.
+
+Unlike ``GenericStruct``s their unpack method accepts a buffer argument with
+its packed body (without the associated header).
+
+How to code a new struct/message
+++++++++++++++++++++++++++++++++
+To implement a new struct, you need to:
+- define a new class which derives from ``GenericStruct``;
+- define class attributes in pack order (with optional initial values) whose
+  type are pyof objects themselves.
+
+Struct example:
+
+.. code:: python
+
+    >>> from pyof.foundation.base import GenericStruct
+    >>> from pyof.foundation.basic_types import UBInt8, UBInt16
+
+    >>> class MyNewStruct(GenericStruct):
+            my_first_attribute = UBInt8(255)
+            my_second_attribute = UBInt16(0)
+            my_third_attribute = UBInt8(255)
+
+which can be used like this:
+
+.. code:: python
+
+    >>> my_struct = MyNewStruct()
+    >>> my_struct.my_second_attribute = 2
+    >>> p = my_struct.pack()
+    >>> print(p)
+    b'\xff\x00\x02\xff'
+
+    >>> my_struct_2 = MyNewStruct()
+    >>> my_struct_2.unpack(p)
+    >>> my_struct_2.my_second_attribute
+    UBInt16(2)
+
+Message example:
+
+.. code:: python
+
+    >>> from pyof.foundation.base import GenericMessage
+    >>> from pyof.foundation.basic_types import UBInt8, UBInt16
+    >>> from pyof.v0x04.common.header import Header
+
+    >>> class MyNewMessage(GenericMessage):
+            header = Header(message_type=255)
+            my_struct_field = MyNewStruct()
+
+which can be used like this:
+
+.. code:: python
+
+    >>> message = MyNewMessage()
+    >>> message.header.xid = 65280
+    >>> message.my_struct_field = my_struct_2
+    >>> message.pack()
+    b'\x04\xff\x00\x0c\x00\x00\xff\x00\xff\x00\x00\xff'
+
+
+
+How to start a new "pyof version"
++++++++++++++++++++++++++++++++++
+
+- Create a new package under ``pyof`` named after the version you are
+  interested in implementing, like ``v0xff`` for example.
+- Create new subpackages for your message categories.
+- Implement your new message as described before.
+
+your file tree should look like this::
+
+  - pyof/
+    + foundation/
+    - v0xff/
+      - my_message_category/
+        __init__.py
+        mynewmessage.py
+      __init__.py
```

Comment:
add this `__init__.py` file before `my_message_category` to improve readability.

Also, if you are doing a "listing", it would be nice to put list markers on all lines.
The same may apply for the first listing from this file.